### ISSUE TITLE
Variable names

### ISSUE THREAD

**Issue by TomDonoghue** [voytekresearch/neurodsp#82]
We don't have super consistent variable naming. With a breaking 1.0 release, should we go through and systematize it all to snake_case (so, for example, things like 'N_samples' -> 'n_samples'). 

Also, some specific cases:
- `filter` overwrites the python `filter` function, which can be really problematic.  
  - Especially because NDSP by default imports all functions directly in local namespace, this squashes filter always, meaning you can never use Python filter, and so has side effects / breaks things
- `psd`, as a function, strikes me as not a great name. Elsewhere, `psd` is used as a variable name, and it sounds like data more than a function
  - Heuristic: function names should be 'verb-y', because they do things.

We also have a bunch of single character variables, but this (in some cases), might be fine for the math-y parts, and also some names, like 'Fs', that are non-compliant, but (I think), are inherited from style in scipy.signal, or similar. Do we want to update those names too?

**Comment by rdgao** [comment]
- could either change the name (use ```filter_ts()```) or don't do the auto imports in __init__, so it doesn't squash unless explicitly done.
- ```compute_spectrum``` ? 
- I'm good with all snake case, including changing Fs to fs.

**Comment by rdgao** [comment]
do we have assignments for this? Are @srcole and I cleaning our respective modules, or somebody wants to just do it wholesale?

**Comment by srcole** [comment]
if you can do `filt` and `spectral` and `sim`, I'll do `timefrequency` and `swm` and `laggedcoherence` and `burst`

**Comment by TomDonoghue** [comment]
For consistency between files - wholesale might be better. 

I'm happy to do a first pass, on a branch, and y'all can scroll through and make any tweaks, if you want?

**Comment by TomDonoghue** [comment]
This is a case where working collaboratively on a branch (as opposed to a fork) is probably useful.

**Comment by TomDonoghue** [comment]
Like, this is a case where it might be good to coordinate. I made a branch with a first pass on 'filt'. Are we going with 'filter_signal', for the function name?

There are a couple more single char variables I would like to update (a, b, w, h) - but not sure best new names. Suggestions?

**Comment by TomDonoghue** [comment]
Some particular suggestions - how do we feel about:
- x -> sig
- Fs -> freqs

(I'll update this as I scroll through.)

**Comment by srcole** [comment]
a, b, w, h are common DSP standards, as used in `scipy.signal.freqz()`
https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.freqz.html

**Comment by srcole** [comment]
Also, I prefer Fs for "sampling frequency". "freqs" looks like "frequencies" to me.
F_s is also a common DSP abbreviation for sampling frequency, as in e.g. https://www.ece.rice.edu/~srs1/files/Lec7.pdf

**Comment by TomDonoghue** [comment]
Yeh... but they still suck (single var). Try updating the filter function right now, to change where how 'a' & 'b' are used... 

'Fs' - ahh yes, of course. So 'fs' then? Or 's_rate' I think is clearer. 

Can we update 'T'?

**Comment by srcole** [comment]
I *really* think they should stay a,b,w,h. Especially cause the only reason they exist is to plot the frequency response, which is generally defined using these variable names

**Comment by voytek** [comment]
Agree with @srcole here. Even if they "suck" they're relatively standard convention, and used in the functions we're calling and emulating from scipy.

**Comment by TomDonoghue** [comment]
'names' is a WIP candidate. All single variable names need manual updating to finish it up (this is why they suck). 

I don't see much reason to be overly attached to scipy conventions, in so far as we can be consistent with but better than, but that's just my vote.

**Comment by rdgao** [comment]
my votes: 
- ```Fs``` -> ```fs```: no capitals to be consistent, and fs is in every single DSP textbook. Perhaps not clear that the s is suppose to be subscript, so I'm fine with ```f_s too```. But ```s_rate``` decouples the "frequency" component of it (as in, Nyquist) and introduces another unnecessary term, which will confuse people when they do go back and read a DSP textbook, as they inevitably will have to at some point.

- keep the filter variable names (a,b,w,h) - again, those are straight out of engineering convention, and would really confuse people if they were to find the equivalent in text if we called it something else. It's not so much to be consistent with scipy but with the engineering conventions. I'd understand if neurodsp is the primary and only reference for people doing DSP, but it is not and it most certainly shouldn't be.

- ```T``` -> ```N_seconds```: ```sig_length``` could indicate sample number, which is not what that variable is. ```N_seconds``` makes it very clear on sight.

**Comment by TomDonoghue** [comment]
Honest question: who is our presumed audience / user base? Because it's not clear it's engineer, which all of this discussion presumes.

**Comment by rdgao** [comment]
that's exactly the point - we're NOT writing for engineers, we're writing for people who may need to consult an engineering textbook at some point with fresh eyes, i.e. probably NOT engineers. The people that can look at the code and be confused about the naming convention are the people we don't need to worry about. Annoying for them, but ultimately not a problem. This is the thought I run into every time I see srate in a dataset, because I know that srate and fs are the same thing. 

It IS a problem if we change it to be more 'user-friendly' and somebody new to DSP just cannot find on the internet or in a book what those terms mean, because some of them don't even have verbal definitions. Are those variable conventions in engineering books 'good'? Arguably not, mostly because they are math symbols and not words. But I don't think we're in the business of fixing engineering convention just quite yet, and I think this falls _more_ within the scope of DSP, instead of best coding practices (which is probably why scipy keeps them as well). 

I don't mind the name change but I am very adamant on this point: we/I do NOT want people to look at our code, think they understand a term with a formal definition because it's in plain English, and use that as a point of reference, instead of scipy or a textbook. We could comment better to point them to the correct resources if that helps.

**Comment by TomDonoghue** [comment]
Yes, of course people can read docstrings, explaining conventions & choices. I'm asking what the conventions should be, and who the module is for. Naming things is very difficult. Being community-driven is the one of the hardest parts of OSS. I think it's worth the discussion. 

There are two types of variables: user facing & internal. I'm not recommending a fundamentally different name (name in terms of what it relates to), anywhere. 

For internals:
- What matter is code maintainability. Single character names are terrible for this. You can't see where they are used, or edit them. So, I'd argue for things like 'a' -> 'a_val' or similar. If you think that is too different or too verbose (and worth making the code harder to maintain), okay - do that. But 'a' & 'b' are literally the worse possible variable names, and I'm unconvinced they are worth keeping as is. 

For public facing:
- Convention & consistency matter, as does clarity. I'm unconvinced by just following what comes before. 

I totally agree with your last point Richard, but have a different view of how it cashes out. The symbol 'Fs' (in math) refers to the concept of sampling rate, yes? They are equivalent symbols, for the same referent? If not, then point taken, but if they are equivalent referents, then I think it's valid to use one or the other. Sampling rate is easier to search for even. If so, math symbol 'Fs'  appears in textbooks, as does the term 'sampling rate', and I don't understand how one or the other somehow decouple us from other resources. Basically: I don't see 'user friendly' as meaning 'less rigorous', and I think people conflate the two much more often than this is actually true. 

At some level, it is a different of opinion on to what degree code should look like math. I think code and math are different. I think it's very different coming at this as an engineer, and coming in from a different entry point, I find 'fs' a much more confusing name. 

Scipy is not PEP8. If we follow scipy, we have chosen a different style guide, which prioritizes math. I disagree with that, for code, and I find scipy code hard to use & read, and it is also hard to maintain - but you can choose to use their style guide, if you prefer.

**Comment by rdgao** [comment]
I'm happy with appending _val to the single letter vars, for the sake of maintainability. 
Rest of them look good to me.
